Cover one more edge case for reverting a dependency to pending

ljacomet opened the pull request:
This change covers one more edge case where the dependency causing the
status revert happens to itself declare an optional dependency to the
reverting module.

Pending state of a dependency is now pushed inside the `ModuleResolveState` which results in less
map lookups for query or modifications.

Requires changes from #6425

melix commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java:
```diff
@@ -291,4 +293,21 @@ VirtualPlatformState getPlatformState() {
         return platformState;
     }
+
+    void removeHardEdge() {
```
:x: Now that the methods are on `ModuleResolveState` I think it can be confusing to have a method named "remove hard edge" or "add hard edge", because this shouldn't deal with edges. Maybe methods like "incrementHardEdgeCount" would make more sense, WDYT?

ljacomet commented:
@melix I have addressed the concerns about naming and at the same time managed to get rid of another code path that was doing a map lookup.
I consider this final though I will wait to send it through CI that your PR on which it is based gets merged.

melix commented:
Let's wait that #6425 is merged into `master`, then rebase onto it.
